cpp/neuralnet/trtbackend.cpp

```diff
@@ -0,0 +1,1441 @@
+    }
+
+    return gpoolConcatLayer;
+  }
+
+  ILayer* applyMaskLayer(ILayer* inputLayer, bool forceFP32 = false) {
+    if(!model->requireExactNNLen) {
+      auto maskLayer = model->network->addElementWise(
+        *inputLayer->getOutput(0), *maskExtractLayer->getOutput(0), ElementWiseOperation::kPROD);
+      auto maskLayerName = string(inputLayer->getName()) + "/mask";
+      maskLayer->setName(maskLayerName.c_str());
+      if(forceFP32) {
+        maskLayer->setPrecision(DataType::kFLOAT);
+      }
+      return maskLayer;
+    } else {
+      return inputLayer;
+    }
+  }
+
+  ILayer* applyCastLayer(ILayer* inputLayer, DataType dataType) {
+    auto castLayer = model->network->addIdentity(*inputLayer->getOutput(0));
+    castLayer->setOutputType(0, dataType);
+    auto castLayerName = string(inputLayer->getName()) + "/cast";
+    castLayer->setName(castLayerName.c_str());
+    return castLayer;
+  }
+};
+
+struct ComputeHandle {
+  bool usingFP16;
+  int modelVersion;
+  ComputeContext* ctx;
+
+  unique_ptr<ICudaEngine> engine;
+  unique_ptr<IExecutionContext> exec;
+  vector<void*> buffers;
+  vector<size_t> bufferBytes;
+  vector<size_t> bufferRowElts;
+  vector<pair<string, Dims>> debugOutputs;
+
+  ComputeHandle(
+    const cudaDeviceProp* prop,
+    ComputeContext* context,
+    const LoadedModel* loadedModel,
+    int maxBatchSize,
+    bool requireExactNNLen) {
+    ctx = context;
+
+    modelVersion = loadedModel->modelDesc.version;
+
+    auto builder = unique_ptr<IBuilder>(createInferBuilder(ctx->trtLogger));
+    if(!builder) {
+      throw StringError("TensorRT backend: failed to create builder");
+    }
+    auto runtime = unique_ptr<IRuntime>(createInferRuntime(ctx->trtLogger));
+    if(!runtime) {
+      throw StringError("TensorRT backend: failed to create runtime");
+    }
+
+    auto config = unique_ptr<IBuilderConfig>(builder->createBuilderConfig());
+    if(!config) {
+      throw StringError("TensorRT backend: failed to create builder config");
+    }
+
+    {
+      usingFP16 = false;
+      if(builder->platformHasFastFp16()) {
+        if(ctx->useFP16Mode == enabled_t::True || ctx->useFP16Mode == enabled_t::Auto) {
+          config->setFlag(BuilderFlag::kFP16);
+          usingFP16 = true;
+        }
+      } else if(ctx->useFP16Mode == enabled_t::True) {
+        throw StringError("CUDA device does not support useFP16=true");
+      }
+      config->setFlag(BuilderFlag::kSTRICT_TYPES);
+    }
+
+    auto network = unique_ptr<INetworkDefinition>(builder->createNetworkV2(0U));
+    if(!network) {
+      throw StringError("TensorRT backend: failed to create network definition");
+    }
+
+    vector<unique_ptr<float[]>> extraWeights;
+    auto modelParser = make_unique<ModelParser>();
+    auto model = modelParser->parse(move(network), loadedModel, ctx->nnXLen, ctx->nnYLen, requireExactNNLen);
+
+    bool saveTimingCache;
+    string timingCacheFile;
+    unique_ptr<ITimingCache> timingCache;
+    {
```

Comment:
All of this timing cache creation - is it potentially inaccurate if this runs at the same time as the GPU is under heavy usage by other parts of the code?

For OpenCL we have this issue, so when it performs tuning of the OpenCL parameters, there is code here: https://github.com/lightvector/KataGo/blob/master/cpp/command/contribute.cpp#L591 that performs tuning for *all* neural net sizes at startup.

Otherwise, we would end up performing some tuning whenever a new size is encountered in the middle of contribute, simultaneously with other threads using the GPU for other games. 

Should we do the same here?